tests/test_schema.py

```diff
@@ -567,3 +590,54 @@ def test_patchset_fail(datadir, patchset_file):
     patchset = json.load(open(datadir.joinpath(patchset_file)))
     with pytest.raises(pyhf.exceptions.InvalidSpecification):
         pyhf.schema.validate(patchset, 'patchset.json')
+
+
+@pytest.fixture
+def refresh_pyhf(monkeypatch):
+    global pyhf
+    modules_to_clear = [name for name in sys.modules if name.split('.')[0] == 'pyhf']
+    for module_name in modules_to_clear:
+        monkeypatch.delitem(sys.modules, module_name)
+    pyhf = importlib.import_module(pyhf.__name__)
+    return pyhf
+
+
+def test_defs_always_cached(
+    socket_disabled,  # noqa: F811
+    refresh_pyhf,  # ensure there is not a pre-existing cache hiding the issue
```

Comment:
I think this could be fixed by monkeypatching the cache directly instead of messing with the importlib.

Reply:
I think if we monkeypatch the cache directly, it makes it impossible to catch the regression in the test.
The behavior the test is meant to assert on is the _initial state of the cache_, i.e. the module-level code. This is only executed exactly once at first import. So we need to re-run the import to be able to catch this behavior.

However, framing it like this makes it clear that the module-level code is really the unit under test here. 
So this should be part of the test body, not a fixture. (it is the act in prepare-act-assert)

Reply:
Fair enough. The test should be the one doing the reimporting in an isolated modules, as you have it. So that looks ok to me (to force the reload).